File: crates/defi-core/src/traits/yield_source.rs
use async_trait::async_trait;

use crate::error::Result;
use crate::types::*;

#[async_trait]
pub trait YieldSource: Send + Sync {
    fn name(&self) -> &str;
    async fn get_yields(&self) -> Result<Vec<YieldInfo>>;
    async fn build_deposit(
        &self,
        pool: &str,
        amount: alloy::primitives::U256,
        recipient: alloy::primitives::Address,
    ) -> Result<DeFiTx>;
    async fn build_withdraw(
        &self,
        pool: &str,
        amount: alloy::primitives::U256,
        recipient: alloy::primitives::Address,
    ) -> Result<DeFiTx>;
}

File: ts/packages/defi-core/src/traits/yield_source.ts
import type { Address } from "viem";
import type { YieldInfo, DeFiTx } from "../types.js";

export interface IYieldSource {
  name(): string;
  getYields(): Promise<YieldInfo[]>;
  buildDeposit(pool: string, amount: bigint, recipient: Address): Promise<DeFiTx>;
  buildWithdraw(pool: string, amount: bigint, recipient: Address): Promise<DeFiTx>;
}
